### PR TITLE
Add catbuffer generator for python #72

### LINKED ISSUE #72: Create catbuffer for Python
Catbuffer generator for Python applications and SDK.

### PR DESCRIPTION
**Pull request by fullcircle23**
Added Catbuffer generator for Python.
Added a new unit test folder for generators.
Added unit tests for generated python code (TransferTransactionBuilder) as a smoke/sanity test.
Added generate_python.sh script.
Updated README.
Closes #72

### CODE REVIEW

**Comment by fullcircle23** [comment]
As the Python generator is based on the Typescript/Java generator, there are enough similarities which cause pylint to report duplicate-code (R0801) in several places. As refactoring would impact all generators which is a big task by itself, the workaround to fix this warning has been to change the code slightly (like renaming a function or inserting a blank line) so that they are no longer similar to pylint. But there are only so many ways to rename a function and to insert another blank line as we are add generators for other languages. Hence, the approach I’ve taken is to set min-similarity-lines=100 in pylintrc file which effectively disables this warning. We should look into refactoring, perhaps after adding a couple more languages to the generators.

**Comment by fboucquez** [comment]
Generators and generated code seem fine. 

I agree there is duplication between the different generators. We may need to have common classes and then delegate to specific helper classes. The alternative is to have some templating files like the mako files. The refactoring is not trivial.

@fullcircle23, are you generating the `resolution_statement` and `state` builders? Typescript is generating them but not Java isn't yet.

I would add `compile.all` and run the test into the `generate_python.sh` script.  I would add more tests if it's possible (we should do the same for java and typescript)

Travis should run the specific` ./scripts/generate_python.sh` instead of `./scripts/generate_all.sh python`. `generate_all.sh `has the problem of incomplete `TransactionBuilderFactory`, `EntityTypeDto`, etc.

**Comment by fullcircle23** [comment]
> * Compile and run the test in generate_python.sh.
> * Add more tests if possible and try using `TransactionBuilderFactory` as an entry point that knows how to create the right builder instance from the binary payload.
> * Double-check that resolution_statements and states are built
> * Update Travis to run generate_python.sh.
> * If possible, publish a pip artifact.

@fboucquez thanks for the review. Please see updates below.
* [DONE] Compile and run the test in generate_python.sh.
* [Separate task] Add more tests if possible and try using TransactionBuilderFactory as an entry point that knows how to create the right builder instance from the binary payload.
* [DONE] Double-check that resolution_statements and states are built
* [DONE] Update Travis to run generate_python.sh.
* [Separate task] If possible, publish a pip artifact.

Some notes about code generation for State schemas:
1. beneficiaryCount
All attributes with key ’name’ with value 'Count' as the suffix such as cosignatoryPublicKeysCount, balancesCount, restrictionsCount and others (total of 17) denote the length/size of a list/array. However beneficiaryCount
(struct HeightActivityBucket) is and the only exception from this convention thus is handled differently from the ‘Counts’ that all share this pattern:
someListCount = uint[8|16|32|64]
someList = array(SomeObject, someListCount)

2. The following generated builders are missing args in their constructor and will be resolved as a separate task.
- AccountRestrictionsInfoBuilder(restrictions1, restrictions2, restrictions3, restrictions4, restrictions5, restrictions6, restrictions7, restrictions8)
- MosaicRestrictionEntryBuilder(addressEntry, globalEntry)
- NamespaceAliasBuilder(mosaicAlias, addressAlias)

**Comment by fullcircle23** [comment]
> Generators and generated code seem fine.
> 
> I agree there is duplication between the different generators. There is duplicated effort every time a language is added or a feature/change is added. We may need to have common classes and then delegate to language-specific helper classes. The alternative is to have some templating files like the mako files. The refactoring is not trivial.
> 
> @fullcircle23, are you generating the `resolution_statement` and `state` builders? Typescript is generating them but Java isn't yet.
> 
> I would add `compile.all` and run the test into the `generate_python.sh` script. The script would then generate, compile and run tests. I would add more tests if it's possible (we should do the same for java and typescript). In the future `generate_python.sh` should deploy a pip artifact like java and typescript (maven central and npm repositories)
> 
> Travis should run the specific` ./scripts/generate_python.sh` instead of `./scripts/generate_all.sh python`. `generate_all.sh` has the problem of incomplete `TransactionBuilderFactory`, `EntityTypeDto`, etc and it doesn't do custom compilation, testing or publishing.

Please see comments below. Thanks.

File: scripts/generate_python.sh [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+rm -rf "${rootDir}/catbuffer/_generated/python"
+rm -rf "$rootDir/build/python/$artifactName"
+
+PYTHONPATH=".:${PYTHONPATH}" python3 "catbuffer/main.py" \
+  --schema catbuffer/schemas/all.cats \
+  --include catbuffer/schemas \
+  --output "catbuffer/_generated" \
+  --generator python \
+  --copyright catbuffer/HEADER.inc
```
Comment (fboucquez) [inline scripts/generate_python.sh]: I would do the other way around. This script should run the compile.all and run the tests. So Travis and devs can test if the generated code is ok

Comment (fboucquez) [reply scripts/generate_python.sh]: Have you changed the all.cats? The current problem we have is that all.cats doesn't include state and resolution statements. All.cat is in the catbuffer repo/submodule which we don't control. We need to manually patch the all.cats locally,  create a new complete all.cats outside catbuffer or call the generator multiple times with different schema inputs.

https://github.com/nemtech/catbuffer/issues/21
https://github.com/nemtech/catbuffer/issues/20

Comment (fullcircle23) [reply scripts/generate_python.sh]: 1) I've left this out to do in a separate task as I just couldn't get test_TransferTransactionBuilder.py to run in travisci due to env / issue with accessing the module/path catbuffer/_generated.
2) Yes. I patched all.cats locally to include generation of schemas under resolution_statement and state.

**Comment by fullcircle23** [comment]
> Approved but please create tasks for the missing parts

Thanks. Created #75 and #76

**Pull request merged**